flink-table/flink-table-common/src/main/java/org/apache/flink/table/annotation/ProcedureHint.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.annotation;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.table.procedures.Procedure;
+import org.apache.flink.table.types.inference.TypeInference;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Repeatable;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * A hint that influences the reflection-based extraction of input types and output types for
+ * constructing the {@link TypeInference} logic of a {@link Procedure}.
+ *
+ * <p>One or more annotations can be declared on top of a {@link Procedure} class or individually
+ * for each {@code call()} method for overloading function signatures. All hint parameters are
+ * optional. If a parameter is not defined, the default reflection-based extraction is used. Hint
+ * parameters defined on top of a {@link Procedure} class are inherited by all {@code call()}
+ * methods. The {@link Procedure} should always return T[], so the {@link DataTypeHint} for output
```

Comment:
Nit: for a dev process, we may first decide on what type to return and then create the corresponding type hint, not vice versa.

{@link DataTypeHint} should always hint output type T since the {@link Procedure} returns T[].

Reply:
Thanks for advide. Since I also want the highlight it's for the case of output type, I would like change it to
```
The {@link DataTypeHint} for the output data type of a {@link Procedure} should always hint the component type of the array returned by {@link Procedure}.
```